castervoice/rules/apps/mouse_grids/gridrainbow.py

```diff
@@ -138,5 +136,19 @@ class RainbowGridRule(MappingRule):
     }
 
 
+def is_rainbow_on():
```

Comment:
This is checked every utterance so we have to be careful of for performance reasons. In the event that it takes longer to process the utterance it delays recognition.

Reply:
mm good point - the question is whether it takes longer than checking the window name though right? Though I presume it's still going to be slower. We could also make this linux only

Reply:
It depends on a number of factors. As I was working on this a couple nights ago here's how I worked out the issue hopefully in a cross-platform wat.

I eliminated the performance concerns completely by unifying the state management with the grids in `navigation`. a few things were needed:

- one function context function to call for all grids. ` function_context=lambda: navigation.is_grid_active("legion")` using a lambda I was able to pass a parameter.

-  re-factored `def kill():` from the individual grids since to  `navigation`

It's important that `mouse_alternates` and the methods that kill the grid are the same file. Principally when `mouse_alternates` stores the process `GRID_PROCESS` and grid mode `MODE` they are stored as global states. if the grid is asked to kill or directly canceled `Kill` is issued with following `wait_for_grid_exit` which polls the process checking to make sure it's terminated. When the process is successfully terminated both `GRID_PROCESS` and `MODE` are set to `None` updating global variables.

By keeping track of the grid process and the mode that requested `GRID_PROCESS` and `MODE` we can use function context to call `is_grid_active` there's no performance penalty here.

`
def is_grid_active(grid):
    global MODE
    if MODE is None:
        return False
    elif MODE == grid:
        return True 
`

[is_grid_active](https://github.com/LexiconCode/Caster/blob/8b2af62fdbc961a425a426c681ef0cec2d8bda6a/castervoice/lib/navigation.py#L27)